### DISCUSSION TITLE
SwAV: Difference to original implementation

### DISCUSSION THREAD

**Discussion by RGring** [KeremTurgutlu/self_supervised#20] [Show and tell]
Hi Kerem,
Thanks for sharing your code and results on imagewang. It was motivating to see positive results on a smaller dataset.

In the last week, I was comparing your swav implementation with the official one of FAIR(https://github.com/facebookresearch/swav). Mainly, because I get much better (and faster) results with yours. Unfortunately, I was not able to find the relevant bit, that makes your implementation better.

Differences, that I already integrated in the original implementation and that didn't make a difference are:
* xresnet
* LabelSmoothingCrossEntropy during fine-tuning
* ranger optimizer
* same training hyperparameters
* same augmentation hyperparameters
* disable extra features like queue, LARC, lr schedule

I was wondering, if you have an idea what else could make the difference? I am running out of ideas...

Thanks in advance!

**Comment by KeremTurgutlu** [comment]
Hi @RGring!

Thanks a lot for conducting these experiments! Indeed, it would be really useful to understand why there might be any differences between FAIR and this repo in terms of performance and/or speed. 

There are couple of questions I would be interested in hearing about to further move on with this investigation :)

- Can you define **better** and **faster**. How do you quantify them, how much better and faster? Can you share experiment logs/results? Are they statistically significant differences?

- I assume you used ImageWang dataset for comparison, is that correct?

 - Can you provide example notebooks, code, etc you used to train both with original FAIR code and this repo, so that anyone interested can investigated and perhaps help us to find the difference.

Appreciate all this work, thanks again!

**Comment by RGring** [comment]
Hi @KeremTurgutlu,
Thanks for your answer.
With better/faster I mean a lower (-->better) loss value with a lower (-->faster) number of epochs. So maybe both are kind of redundant.

I started out with ImageWang, but continued later with my targeted dataset. Though, I remember that I observed similar differences with the ImageWang dataset.

I am still running experiments. But it seems to be a mix of different things. I can share my insight, once I can say for sure.

I would be curious, what are your reasons/motivations for the following design choices.

- xresnet as encoder
- ranger optimizer
- augmentation library kornia
- AdaptiveConcatPool2d instead of AdaptiveAvgPool2d after encoder.

Best regards!

**Comment by KeremTurgutlu** [comment]
Hi @RGring,

Most of the hyperparameter choices, **xresnet, ranger,  and AdaptiveConcatPool2d**, are made based on previous ImageWang experiments conducted by other people to make a fair comparison with their scores. Also, these selections are adopted defaults set by fastai in general. I used kornia since it allows augmenting batch of images instead individual images, so that might be one of the reasons for faster training? One could use torchvision.transforms as well.

I am very excited to hear about your conclusions and insights once you are done experimenting. Maybe imagewang, and your dataset's nature is more suitable for the implementation we have here compared to FAIR's. It would be interesting to run comparisons on datasets they mentioned in their paper, but that would require time and compute. Perhaps, that can be done later by me :)

**Comment by RGring** [comment]
Hi Kerem,
I integrated bits of your implementation in the original swAV implementation to find out what makes the difference. The main improvement are due to the **xResnet** architecture and the **AdaptiveConcatPool2d** (instead of the AdaptiveAvgPool2d) after the backbone. Both design choices lead to better final accuracies. Furthermore representations are learned much faster, i.e. with fewer number of epochs.
In the attached figure, you can see avg top-1 accuracy for different epochs and different models (experiments were performed on my own dataset, 9 classes, ~9000 images):

- resnet: resnet-34 + AdaptiveAvgPool2d
- resnet_concat_pool: resnet-34 + AdaptiveConcatPool2d
- resnet_tweaks: xresnet-34 + AdaptiveConcatPool2d

![image](https://user-images.githubusercontent.com/19685020/104329263-07d03500-54ed-11eb-8cab-128695f2e228.png)

First, I also thought that the kornia-library. But I found out, that the augmentation function RandomResizedCrop from kornia is not performing the same as the one of the albumentation library. So I had to adapt the hyperparameters, here.

I hope that helps. :).

**Comment by KeremTurgutlu** [comment]
This is fantastic work, thank you!

**Comment by RGring** [comment]
Thanks. The good thing about the success of contrastive learning is, that now we can get good pretrained weights for any crazy architectures. :)

**Comment by oguiza** [comment]
Hi @KeremTurgutlu,
I'd like to thank you for putting together this implementation of SwAV. It's great! I also use fastai and am now working on an implementation for time series using yours as a basis.
And thanks a lot @RGring  for your analysis. It's an excellent way to better understand the performance of a model.

There's something I've read from the primary author of SwAV that might be of interest to you: 
"Yes I would recommend adjusting the number of prototypes to your custom dataset. For example if your dataset has 10 classes, you can try using --nmb_prototypes 100" (you can read the context here: https://github.com/facebookresearch/swav/issues/37). 

In the paper, they use 3k prototypes for Imagenet, but your datasets have far fewer classes (9 or 20 I think). So the # prototypes should probably be way lower (i.e. 10x n classes). Have you tested SwAV with a different number of prototypes?

**Comment by RGring** [comment]
@oguiza Yes, I tried 50, 300, 1000 and 3000 number of prototypes and unfortunately there was no significant difference in the final performance. But I am curious, if you experience differently.

**Comment by oguiza** [comment]
Thanks @RGring for your reply. I see you've done a pretty thorough analysis!
It's interesting. My limited experience so far is very similar. I have not been able to see much difference, although it's a bit early to say.

**Comment by kai-tub** [comment]
I would also like to thank both of you @RGring  and @KeremTurgutlu for this insightful discussion!
I am also taking a closer look at the code and have one remaining question regarding the `kornia` library.

> I used kornia since it allows augmenting batch of images instead individual images, so that might be one of the reasons for faster training? One could use torchvision.transforms as well.

Is there any particular reason for why you've chosen `kornia` over the _default_ fastai transforms?

Fastai provides transformation such as [`FlipItem`](https://docs.fast.ai/vision.augment.html#FlipItem), [`RandomResizedCrop`](https://docs.fast.ai/vision.augment.html#RandomResizedCrop) and [`grayscale`](https://docs.fast.ai/vision.augment.html#grayscale).
All of which can also be applied to batches on the GPU.

Thanks in advance! :)

**Comment by KeremTurgutlu** [comment]
Hey @kai-tub!

I tried to follow augmentations that are used in SimCLR paper as well as in the open source implementations of BYOL. Few of them didn’t have direct correspondence in fastai, and it seemed like a better idea to use what has been already tested by others. Now, that we know the code works and gives good results we can use any transformation library as long as it supports batch transforms since it’s faster.

**Comment by kai-tub** [comment]
Thanks for the quick response!
Ok, I was just wondering if I was missing a _benefit_ of `kornia` but then I am going to stick to the classic `fastai` augmentations for my custom transforms. :)
